engines/sci/graphics/palette32.cpp

```diff
@@ -21,20 +21,156 @@
  */
 
 #include "common/file.h"
+#include "common/system.h"
+#include "graphics/palette.h"
 
 #include "sci/sci.h"
+#include "sci/resource.h"
 #include "sci/graphics/palette32.h"
+#include "sci/graphics/screen.h"
 
 namespace Sci {
 	
 GfxPalette32::GfxPalette32(ResourceManager *resMan, GfxScreen *screen)
-	: GfxPalette(resMan, screen), _clutTable(0), _cyclers(), _cycleMap() {}
+	: GfxPalette(resMan, screen),
+	_clutTable(nullptr),
+	// Palette cycling
+	_cyclers(), _cycleMap(),
+	// Palette varying
+	_varyTime(0), _varyDirection(0), _varyTargetPercent(0),
+	_varyTargetPalette(nullptr), _varyStartPalette(nullptr),
+	_varyFromColor(0), _varyToColor(255), _varyNumTimesPaused(0),
+	_varyPercent(_varyTargetPercent), _varyLastTick(0),
+	// Palette versioning
+	_version(1), _needsUpdate(false) {
+		// TODO: Sci::Palette should have a default constructor that does
+		// all of this
+		for (int color = 0; color < ARRAYSIZE(_sysPalette.colors); color++) {
+			_sourcePalette.colors[color] = _nextPalette.colors[color] = _sysPalette.colors[color];
+			// TODO: Intensity and mapping are not used by SCI32 but are used by
+			// other parts of the graphic engine right now. Replace those parts.
+			_sourcePalette.intensity[color] = _nextPalette.intensity[color] = 100;
+			_sourcePalette.mapping[color] = _nextPalette.mapping[color] = color;
+		}
+
+		for (int i = 0; i < ARRAYSIZE(_fadeTable); ++i) {
+			_fadeTable[i] = 100;
+		}
+
+		// TODO: In SCI engine, GfxPalette (PaletteMgr) constructor handles
+		// loading the default palette, but ScummVM already loads it from
+		// somewhere else, so should ScummVM be updated to match SCI or should
+		// it be left the way it is already?
+		assert(kernelSetFromResource(999, true));
+	}
 
 GfxPalette32::~GfxPalette32() {
 	unloadClut();
+	varyOff();
 	cycleAllOff();
 }
 
+void GfxPalette32::submit(Palette &palette) {
+	// TODO: This code won’t do much unless/until submit can
+	// receive HunkPalettes in addition to SOLPalettes. In SCI engine,
+	// SOLPalettes do not have timestamp/version fields, only
+	// HunkPalettes do, but in ScummVM there is only one type of palette.
+	if (palette.timestamp == _version) {
+		return;
+	}
+
+	// TODO: Verify this is doing a copy.
+	Palette oldSourcePalette(_sourcePalette);
+
+	// TODO: In SCI32 this is actually calling SOLPalette::merge,
+	// so there should probably be a merge function on Sci::Palette
+	// instead of doing it inline here like this.
+	for (int i = 0; i < ARRAYSIZE(_sourcePalette.colors); ++i) {
+		if (palette.colors[i].used) {
+			_sourcePalette.colors[i] = palette.colors[i];
+		}
+	}
+
+	if (!_needsUpdate && _sourcePalette != oldSourcePalette) {
+		++_version;
+		_needsUpdate = true;
+	}
+
+	// TODO: Technically this information is supposed to be
+	// persisted through a HunkPalette object; right now it
+	// would just be lost once the temporary palette was destroyed.
+	// palette.timestamp = _version;
```

Comment:
There aren’t hunk segments stored in save games—at least, not from `PaletteMgr::SaveRestore`. The persistence seems like it is probably just in-memory… perhaps the way Sierra chose to optimise palette resource loads because of how multiple palettes get merged into a single source palette?

The code that does it seemed pretty crazy at the time I looked at it, so I didn’t fully RE to figure out exactly how it was working. There is some pointer at `HunkPalette+0` which is set in `HunkPalette::Init` (this value always seems to be the last allocated handle from `MemoryMgr` immediately after a `ResourceMgr::load` call to load a palette), plus a value at `HunkPalette+4` which comes from `[HunkPalette+0]+0a`. When `HunkPalette::SetVersion` is called and `HunkPalette+4` is not zero, some arithmetic is performed to find the version number again (`HunkPalette+0 + 0x0d + 2 * HunkPalette+4 + 0x12`), and it’s compared to the `HunkPalette::version` field. `MsgMgr::Fatal` is called if they don’t match, otherwise the `newVersion` argument is put to both `HunkPalette::version` and that computed memory handle offset.